feat!(protocol): make payments for all record types

joshuef opened the pull request:
## Description

<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 04 Sep 23 04:58 UTC
This pull request includes multiple file diffs with various changes. Here is a summary of the changes:

- In the `api.rs` file, the `validate_and_store_record` function call was modified to remove the boolean argument.
- The `sn_protocol/src/messages/mod.rs` file has changes related to the `ChunkWithPayment`, `Chunk`, and `ReplicatedData` structs.
- The `log_markers.rs` file has changes related to the `Marker` enum, where new variants were added.
- The `error.rs` file has changes related to the renaming of the `ChunkNotStored` error variant and updating the associated error message.
- The `sn_protocol/src/storage/mod.rs` file has changes related to import modifications and removals.
- The `sn_protocol/src/storage/chunks.rs` file has changes related to import removals and the removal of the `ChunkWithPayment` struct.
- The `lib.rs` file has changes related to import additions and improvements to existing methods.
- The `Client` implementation in the `lib.rs` file has changes related to the `Store chunk` method, including the removal of the `ChunkWithPayment` struct.
- The `header.rs` file has changes related to the `RecordKind` enum and the `RecordHeader` struct, including new variants, serialize and deserialize trait implementations, and additional methods and test cases.
- The `replication.rs` file has changes related to the modification of the `validate_and_store_record` function call by removing the `false` argument.

These changes indicate modifications to various parts of the codebase, including function calls, imports, enums, structs, error messages, and test cases. Let me know if you need further assistance.
<!-- reviewpad:summarize:end -->

connec commented:
I've drafted this for now in favour of #699 which is a much easier win, albeit less idealistic.

joshuef committed:
feat!(protocol): make payments for all record types

joshuef committed:
chore: se/derialize for PrettyPrintRecordKey

grumbach commented on sn_node/src/put_validation.rs:
```diff
@@ -30,48 +29,135 @@ use std::collections::{BTreeSet, HashSet};
+            return Err(ProtocolError::PaymentProofInsufficientAmount {
+                paid: Token::zero(),
+                expected: Token::zero(),
+            });
```
Here this error message might be confusing, why don't we create another error type like `RecordMissingPayment`?

joshuef replied:
i've dropped this entirely with the dbcspendwithpayment removal 👍

RolandSherwin commented on sn_protocol/src/storage/header.rs:
```diff
@@ -123,6 +130,11 @@ mod tests {
+            kind: RecordKind::ChunkWithPayment,
+        }
+        .try_serialize()?;
+        assert_eq!(record_with_payment.len(), RecordHeader::SIZE);
```
Can we also add the check for `RegisterWithPayment`. We had some undefined behaviour with a specific variant earlier if I'm not wrong.

RolandSherwin commented on sn_node/src/put_validation.rs:
```diff
@@ -30,48 +29,121 @@ use std::collections::{BTreeSet, HashSet};
+
+                    if already_exists {
+                        return Ok(CmdOk::DataAlreadyPresent);
+                    }
```
Should we return early here? If a `SignedSpend` already exists locally, then we might have to make sure that the new PUt is not a double spend?
I think it is performed inside `validate_and_store_spends`.

joshuef replied:
ahaaa, okay good spot

joshuef committed:
test: add RegisterWithSpend header validation

joshuef committed:
chore(node): refactor and extract out spend validation.

We also avoid exiting early if spends exist as we may have a double spend on our hands
